jupyter_server/services/kernels/connection/channels.py

```diff
@@ -268,6 +282,63 @@ def nudge(count):
         future.add_done_callback(finish)
         return _ensure_future(future)
 
+    def wait_for_iopub_welcome(self):
+        """Wait for an iopub welcome message
+        Since protocol_version >= 5.4
+        """
+        iopub_channel = self.channels["iopub"]
+        iopub_future: Future = Future()
+
+        def on_iopub(msg):
+            """Handle iopub welcome message"""
+            idents, msg = self.session.feed_identities(msg)
+            try:
+                msg = self.session.deserialize(msg)
+            except BaseException:
+                self.log.error("Bad Iopub message", exc_info=True)
+                iopub_channel.stop_on_recv()
+                iopub_future.set_result(None)
+            else:
+                msg_type = msg["header"]["msg_type"]
+                if msg_type == "iopub_welcome":
+                    self.log.debug("Iopub welcome message received: %s", self.kernel_id)
+                    iopub_channel.stop_on_recv()
+                    iopub_future.set_result(msg)
+                else:
+                    self.log.error(
+                        "Iopub welcome message not received, receiving %s instead", msg_type
+                    )
+                    iopub_channel.stop_on_recv()
+                    iopub_future.set_result(None)
+
+        iopub_channel.on_recv(on_iopub)
+        loop = IOLoop.current()
+
+        def give_up():
+            """Don't wait forever for the kernel"""
+            if iopub_future.done():
+                return
+            self.log.error("Timeout waiting for IOPub welcome message from %s", self.kernel_id)
+            iopub_channel.stop_on_recv()
+            iopub_future.set_result(None)
+
+        # Resolve with a timeout if we get no response
+        timeout_future = gen.with_timeout(loop.time() + self.kernel_info_timeout, iopub_future)
```

Comment:
Not sure if we should use `kernel_info_timeout` or something else for the timeout?